File: apps/workflow-studio/src/features/workflow-editor/types/aiChat.ts
/**
 * Types for the AI Chat sidebar feature.
 */

export interface AIChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: Date;
  operations?: AIResponsePayload;
}

export interface AIChatRequest {
  message: string;
  session_id: string;
  workflow_context: WorkflowContextPayload | null;
  conversation_history: { role: 'user' | 'assistant'; content: string }[];
  mode_hint: 'auto' | 'generate' | 'modify' | 'explain' | 'fix';
}

interface WorkflowContextPayload {
  name: string;
  nodes: Array<{ name: string; type: string; parameters: Record<string, unknown> }>;
  connections: Array<{
    source_node: string;
    target_node: string;
    source_output: string;
    target_input: string;
  }>;
}

export interface AIResponsePayload {
  mode: 'full_workflow' | 'incremental' | 'explanation';
  workflow: {
    name: string;
    nodes: Array<{ name: string; type: string; parameters: Record<string, unknown> }>;
    connections: Array<{
      source_node: string;
      target_node: string;
      source_output?: string;
      target_input?: string;
    }>;
  } | null;
  operations: AIOperation[] | null;
  summary: string;
}

export type AIOperation =
  | { op: 'addNode'; type: string; name: string; parameters: Record<string, unknown>; connect_after?: string }
  | { op: 'updateNode'; name: string; parameters: Record<string, unknown> }
  | { op: 'removeNode'; name: string }
  | { op: 'addConnection'; source_node: string; target_node: string; source_output?: string; target_input?: string }
  | { op: 'removeConnection'; source_node: string; target_node: string };

File: apps/workflow-engine/src/schemas/ai_chat.py
"""Schemas for AI chat endpoint."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel


class AIChatMessage(BaseModel):
    """A single message in the conversation history."""

    role: Literal["user", "assistant"]
    content: str


class WorkflowContext(BaseModel):
    """Current workflow state sent from the frontend."""

    name: str
    nodes: list[dict[str, Any]]
    connections: list[dict[str, Any]]


class AIChatRequest(BaseModel):
    """Request body for POST /api/ai/chat."""

    message: str
    session_id: str | None = None
    workflow_context: WorkflowContext | None = None
    conversation_history: list[AIChatMessage] = []
    mode_hint: Literal["auto", "generate", "modify", "explain", "fix"] = "auto"
    test_input: dict[str, Any] | None = None
